beps/0006-scaffolder-action-rollback/README.md

```diff
@@ -1,84 +1,303 @@
+    CP2-->>NunjucksWorkflowRunner: Rollback Complete
+    NunjucksWorkflowRunner->>CP1: Rollback (Repository)
+    CP1-->>NunjucksWorkflowRunner: Rollback Complete
+
+    NunjucksWorkflowRunner-->>Backstage UI: Template Execution Failed
+    Backstage UI-->>User: Display Error & Rollback Status
+```
 
 ## Design Details
 
-We are going to introduce an extra function in action API which will look like:
+### Extending Checkpoint API for Rollback
+
+The checkpoint API will be extended to include an optional rollback function that can utilize the data saved by the checkpoint:
+
+```typescript
+/**
+ * Extended Checkpoint API with rollback support
+ */
+async checkpoint<T>(
+  key: string,
+  fn: () => Promise<T>,
+  rollback?: (data: T) => Promise<void>
+): Promise<T>;
+```
+
+This allows the checkpoint to save data during forward execution that can be used to properly rollback the operation if needed.
+
+Example usage:
+
+```typescript
+// Repository creation with rollback
+await ctx.checkpoint(
+  'repository-created',
+  async () => {
+    const { owner, repo } = ctx.input;
+    const repository = await githubClient.repos.create({ owner, repo });
+    return {
+      repositoryId: repository.id,
+      owner,
+      repo,
+    };
+  },
+  async data => {
+    // Rollback function using saved checkpoint data
+    const { owner, repo } = data;
+    await githubClient.repos.delete({ owner, repo });
+  },
+);
+```
+
+The checkpoint state will be extended to include a new status for rolled back operations. This could include new statuses for failed and rolled back operations.
+
+```typescript
+type CheckpointState =
+  | {
+      status: 'failed';
+      reason: string;
+    }
+  | {
+      status: 'success';
+      value: JsonValue;
+    }
+  | {
+      status: 'rolledback';
+      originalValue: JsonValue;
+    }
+  | {
+      status: 'rollback-failed';
+      originalValue: JsonValue;
+      reason: string;
+    };
+```
+
+This enhancement allows us to identify checkpoints that failed to rollback for potential manual intervention, thus tracking the complete lifecycle of a checkpoint (creation → potential rollback).
+
+### WorkflowRunner Enhancements
+
+The `NunjucksWorkflowRunner` will be modified to track successfully completed checkpoints with their rollback functions. When a scaffolder task fails, the system will:
+
+1. Query the database for completed checkpoints associated with the task
+2. Filter for checkpoints that have rollback functions defined
+3. Execute these rollback functions in reverse order
+
+The implementation will ensure that:
+
+- Rollbacks are executed in proper LIFO order across all actions
+- Each rollback operation is properly logged and tracked
+- **Failures in one rollback don't prevent other rollbacks from being attempted**
+
+```typescript
+// Pseudo-code for rollback execution
+for (const checkpoint of completedCheckpointsWithRollback.reverse()) {
+  try {
+    await checkpoint.rollback(checkpoint.data);
```

Comment:
```suggestion
    await checkpoint.rollback?.(checkpoint.data);
```

Reply:
Updated in https://github.com/backstage/backstage/pull/29196/commits/24c7b7093bb2ddb3fd5a7f03533e5ebd34d72351